File: PHP/Node.php
<?php 

declare(strict_types=1);

class Node {

public static int $nextId = 0;
public int $id;
public int $value;
public $left;
public $right;

function getThis(){
    return $this;
}

function __construct(int $number){
    $this->value = $number;
    $this->id = self::setId();

}


function getNode(int $number){
    return new Node($number);
}

static function setId(){
    self::$nextId++;
    return self::$nextId;


}


// Check if node right exists
function hasRight(){
    if (isset($this->right->value)){
        echo("existe");
        return true;
    }
    echo("não existe");
    return false;
}

// Check if node left exists
function hasLeft(){
    if (isset($this->left->value)){
        return true;
    }
    return false;
}



function isNextNotEmpty(Node $nodeParent, Node $newNode){
    if ($newNode > $nodeParent){
        if (!isset($nodeParent->right)){
            $nodeParent->right = $newNode;
            return false;
        }
        return true;
    }
    if (!isset($nodeParent->left)){
        $nodeParent->left = $newNode;
        return false;
    }
    return true;

}


function addNode(int $number){
    // Note to be added at the end of the tree
    $newNode = new Node($number);
    $auxNode = $this;

    
    
    //Checking right node of the root
    if ($number > $this->value){
        if (!$this->hasRight()){
            $this->right = $newNode;
            return $newNode;
        }


        $parentNode = $this->right;
        $continue = true;
        while($continue){
            $nextLeft = $parentNode->left;
            $nextRight = $parentNode->right;
            if ($newNode->value > $nextLeft->value){
                $continue = $this->isNextNotEmpty($nextLeft,$newNode);
            }
            if ($newNode->value < $nextLeft->value){
                $continue = $this->isNextNotEmpty($nextRight,$newNode);
            }
        }

    }

    //Checking left node of the root
    if ($number < $this->value){
        if (!$this->hasLeft()){
            $this->left = $newNode;
            return $newNode;
        }

        $parentNode = $this->left;
        $continue = true;
        while($continue){
            $nextLeft = $parentNode->left;
            $nextRight = $parentNode->right;
            if ($newNode->value > $nextLeft->value && ($nextLeft->value != null)){
                $continue = $this->isNextNotEmpty($nextLeft,$newNode);
            }
            if ($newNode->value < $nextLeft->value && ($nextRight->value != null)){
                $continue = $this->isNextNotEmpty($nextRight,$newNode);
            }
            
        }
    }

}


}

File: Node.java
public class Node {

    public static int nextId = 0;
    public int id;
    public int value;
    public Node left;
    public Node right;

    public Node getThis(){
        return this;
    }

    public Node(int number){
        this.value = number;
        this.id = setId();

    }


    public Node getNode(int number){
        return new Node(number);
    }

    private static int setId(){
        nextId++;
        return nextId;


    }


    // Check if node right exists
    public boolean hasRight(){
        if (this.right != null){
            System.out.println("existe");
            return true;
        }
            System.out.println("não existe");
        return false;
    }

    // Check if node left exists
    public boolean hasLeft(){
        if (this.left != null){
            return true;
        }
        return false;
    }

    // Add a node using a number
    public void addNode(int number){
        // Note to be added at the end of the tree
        Node newNode = new Node(number);
        System.out.println("Antes do if");

        if (number > this.value){
            if (!this.hasRight()){
                this.right = newNode;
                return;
            }
            this.throwRight(newNode);
            return;
        }
        if (number < this.value){
            if (!this.hasLeft()){
                this.left = newNode;
                return;
            }
            this.throwLeft(newNode);
            return;
        }
    }


    // Add a node using a node
     public void addNode(Node newNode){
        // Note to be added at the end of the tree
        int number = newNode.value;

        if (number > this.value){
            if (!this.hasRight()){
                System.out.println("Saiuuu");
                this.right = newNode;
                return;
            }
            this.throwRight(newNode);
            return;
        }
        if (number < this.value){
            if (!this.hasLeft()){
                this.left = newNode;
                return;
            }
            this.throwLeft(newNode);
            return;
        }
    }

    public void throwLeft (Node newNode){
        
        Node parentNode = this.left;
        while((parentNode.hasLeft() && parentNode.hasRight())){
            Node nextLeft = parentNode.left;
            Node nextRight = parentNode.right;
            if (newNode.value > nextLeft.value){
                nextLeft.addNode(newNode.value);
            }
            if (newNode.value < nextLeft.value){

            }
        }
        if (newNode.value > parentNode.value){
            parentNode.right = newNode;
            return;
        } 
        parentNode.left = newNode;
        
    }

    public void throwRight (Node newNode){

        Node parentNode = this.right;
        while((parentNode.hasLeft() && parentNode.hasRight())){
            System.out.println("Testeeeeee");
            Node nextLeft = parentNode.left;
            Node nextRight = parentNode.right;
            if (newNode.value > nextLeft.value){
                nextLeft.addNode(newNode.value);
            }
            if (newNode.value < nextLeft.value){
                System.out.println("Ué");
            }
        }
        if (newNode.value > parentNode.value){
            parentNode.right = newNode;
            return;
        } 
        parentNode.left = newNode;
        

    }
}
